### PR TITLE
Feature/file compression

### PR DESCRIPTION
**Pull request by jemten**
### This PR fixes:

- see commit message

### How to test:

- Automatic and continuous test pass

### Expected outcome:
- Installation, unit and integration tests pass

### Review:
- [x] Code review
- [x] New code is executed and covered by tests
- [x] Tests pass

### CODE REVIEW

File: lib/MIP/Parse/Gender.pm [hunk 1]
```diff
@@ -494,18 +491,16 @@ sub update_gender_info {
...
-sub _get_file_gzipped_status {
+sub _get_file_read_commands {
 
-## Function : Check gzipped status of file path to choose correct cat binary (cat or zcat). Also prepend stream character.
-## Returns  : $cmd
-## Arguments: $cmds_cat_ref => Command array for cat {REF}
-##          : $file_path    => Fastq File path
+## Function : Check gzipped status of file path to choose correct cat binary (cat or gzip). Also prepend stream character.
+## Returns  : @read_cmds
+## Arguments: $file_path    => Fastq File path
```
Comment (henrikstranneheim) [inline lib/MIP/Parse/Gender.pm]:
```suggestion
## Arguments: $file_path => File path to check status for
```

File: lib/MIP/Parse/Gender.pm [hunk 2]
```diff
@@ -514,38 +509,39 @@ sub _get_file_gzipped_status {
...
-            strict_type => 1,
-        },
     };
 
     check( $tmpl, $arg_href, 1 ) or croak q{Could not parse arguments!};
 
-    use MIP::Parse::File qw{ parse_file_suffix };
+    use MIP::File::Path qw{ check_gzipped };
+    use MIP::Program::Gzip qw{ gzip };
+    use MIP::Gnu::Coreutils qw{ gnu_cat };
```
Comment (henrikstranneheim) [inline lib/MIP/Parse/Gender.pm]: Order AB

File: lib/MIP/Program/Gzip.pm [hunk 3]
```diff
@@ -39,12 +36,12 @@ sub gzip {
 ## Arguments: $decompress             => Decompress
 ##          : $filehandle             => Filehandle to write to (scalar undefined)
 ##          : $force                  => Force overwrite of output file and compress links
-##          : $infile_path            => Infile path
+##          : $infile_paths_ref       => Infile pathis
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Gzip.pm]:
```suggestion
##          : $infile_paths_ref       => Infile path
```

File: lib/MIP/Set/File.pm [hunk 4]
```diff
@@ -54,7 +54,7 @@ sub set_file_compression_features {
 
     use MIP::File::Path qw{ check_gzipped };
 
-    my $read_file_command = q{zcat};
+    my $read_file_command = q{gzip -d -c};
```
Comment (henrikstranneheim) [inline lib/MIP/Set/File.pm]: This was the one you had to hardcode right?

File: t/decompress_files.t [hunk 5]
```diff
@@ -58,44 +58,40 @@ diag(   q{Test decompress_file from Decompression.pm v}
...
-
 ## Given a file to decompress, when method gzip unzip and tar
-my $outfile_path  = q{a_file_path.gz};
-my $reference_dir = q{a_reference_dir};
-my @programs      = qw{ gzip unzip tar };
+my $file_paths_ref = [q{a_file_path.gz}];
+my $file_path      = q{a_file_path.gz};
+my $outdir_path    = q{a_outdir};
+my $outfile_path   = q{a_outfile};
+my @programs       = qw{ gzip unzip tar };
```
Comment (henrikstranneheim) [inline t/decompress_files.t]:
```suggestion
my @programs       = qw{ gzip tar unzip };
```

File: t/decompress_files.t [hunk 6]
```diff
@@ -58,44 +58,40 @@ diag(   q{Test decompress_file from Decompression.pm v}
...
-    decompress_file(
+    my @decompress_commands = decompress_files(
         {
-            filehandle         => $filehandle,
-            outdir_path        => $reference_dir,
-            outfile_path       => $outfile_path,
-            decompress_program => $program,
+            outdir_path  => $outdir_path,
+            outfile_path => $outfile_path,
+            file_path    => $file_path,
```
Comment (henrikstranneheim) [inline t/decompress_files.t]: order AB

File: t/decompress_files.t [hunk 7]
```diff
@@ -58,44 +58,40 @@ diag(   q{Test decompress_file from Decompression.pm v}
...
-    ok( $returned_base_command, qq{Wrote $program command for decompression} );
-
-}
+## Given missing file_paths for tar
+trap {
+    decompress_files(
+        {
+            outdir_path    => $outdir_path,
+            outfile_path   => $outfile_path,
+            file_paths_ref => $file_paths_ref,
```
Comment (henrikstranneheim) [inline t/decompress_files.t]: order AB

File: t/parse_file_suffix.t [hunk 8]
```diff
@@ -84,4 +84,17 @@ $outfile_path_no_suffix = parse_file_suffix(
...
+$compressed_file_name = catfile(qw{ a dir files.tar.gz });
+
+$outfile_path_no_suffix = parse_file_suffix(
+    {
+        file_name   => $compressed_file_name,
+        file_suffix => $DOT . q{gz},
+    }
+);
+
+$file_name = catfile(qw{ a dir files.tar });
```
Comment (henrikstranneheim) [inline t/parse_file_suffix.t]: Nice

File: lib/MIP/Set/File.pm (hunk 4 above)
Comment (jemten) [reply lib/MIP/Set/File.pm]: Yeah. Some of the tests were updated with the hardcoded command as well

File: t/parse_file_suffix.t (hunk 8 above)
Comment (jemten) [reply t/parse_file_suffix.t]: thanks!

File: lib/MIP/File/Decompression.pm [hunk 9]
```diff
@@ -28,103 +27,111 @@ BEGIN {
...
     my $tmpl = {
-        decompress_program => { store => \$decompress_program, strict_type => 1, },
-        filehandle  => { defined => 1, required => 1, store => \$filehandle, },
+        file_path => {
+            defined     => 1,
+            store       => \$file_path,
+            strict_type => 1,
+        },
+        file_paths_ref => {
+            default     => $arg_href->{file_paths_ref} ||= [ $arg_href->{file_path} ],
```
[the remark this answers is not in the document]
Comment (jemten) [reply lib/MIP/File/Decompression.pm]: I'm beginning to like it

**Comment by henrikstranneheim** [comment]
⭐️ I tested the tests locally on the MAc and they all work now!! Nicely done!

**Pull request merged**